src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -1558,73 +1638,141 @@ export function softwareRasterize<T extends Dimensionality>(
+  // writes d, 0, 0, 1 so write the same.
+  const depthFill = depthTexture
+    ? ', 0, 0, 1'
+    : isDepthTextureFormat(texture.format)
+    ? '.r, 0, 0, 1'
+    : '';
 
   // Compute the amount we need to multiply the unitQuad by get the
   // derivatives we want.
   const uMult = (ddx * renderTarget.width) / texture.width;
   const vMult = (ddy * renderTarget.height) / texture.height;
 
-  const offsetWGSL = offset ? `, vec2i(${offset[0]},${offset[1]})` : '';
+  const numLayers = texture.depthOrArrayLayers / (isCubeViewDimension(viewDescriptor) ? 6 : 1);
+  const arrayIndexWGSL =
+    arrayIndexType === 'i'
+      ? `, arrayIndexAsI32(v)`
+      : arrayIndexType === 'u'
+      ? ', arrayIndexAsU32(v)'
+      : '';
+  const offsetWGSL = offset ? `, ${wgslExprT(offset)}` : '';
+  const uvStartWGSL = wgslExprT(extendWithOnes(uvwStart, 3));
+  const uvMultWGSL = wgslExprT([uMult, vMult, 1.0]);
 
   const code = `
 struct InOut {
   @builtin(position) pos: vec4f,
-  @location(0) uv: vec2f,
+  @location(0) uvw: vec3f,
 };
 
 @vertex fn vs(@builtin(vertex_index) vertex_index : u32) -> InOut {
   let positions = array(
-    vec2f(-1,  1), vec2f( 1,  1),
-    vec2f(-1, -1), vec2f( 1, -1),
+    vec3f(-1,  1, 0.5), vec3f( 1,  1, 1  ),
+    vec3f(-1, -1,   0), vec3f( 1, -1, 0.5),
   );
   let pos = positions[vertex_index];
   return InOut(
-    vec4f(pos, 0, 1),
-    (pos * 0.5 + 0.5) * vec2f(${uMult}, ${vMult}) + vec2f(${uvwStart[0]}, ${uvwStart[1]}),
+    vec4f(pos, 1),
+    (pos * vec3f(0.5, 0.5, 1.0) + vec3f(0.5, 0.5, 0.0)) * ${uvMultWGSL} + ${uvStartWGSL},
   );
 }
 
-@group(0) @binding(0) var          T    : texture_2d<f32>;
+@group(0) @binding(0) var          T    : ${textureWGSL};
 @group(0) @binding(1) var          S    : sampler;
 
+// returns -1 to depthOrArrayLayers inclusive
+fn arrayIndexAsI32(v: InOut) -> i32 {
+  return (i32(v.pos.x) + i32(v.pos.y)) % ${numLayers + 2} - 1;
+}
+
+// returns 0 to depthOrArrayLayers inclusive
+fn arrayIndexAsU32(v: InOut) -> u32 {
+  return (u32(v.pos.x) + u32(v.pos.y)) % ${numLayers + 1};
+}
+
 @fragment fn fs(v: InOut) -> @location(0) vec4f {
-  return textureSample(T, S, v.uv${offsetWGSL});
+  return vec4f(textureSample(T, S, ${coordWGSL}${arrayIndexWGSL}${offsetWGSL})${depthFill});
 }
 `;
 
-  const shaderModule = device.createShaderModule({ code });
+  const shaderModules = s_deviceToDrawTextureShaderModules.get(device) ?? new Map();
+  s_deviceToDrawTextureShaderModules.set(device, shaderModules);
+  const shaderModule = shaderModules.get(code) ?? device.createShaderModule({ code });
+  shaderModules.set(code, shaderModule);
 
-  const pipeline = device.createRenderPipeline({
-    layout: 'auto',
-    vertex: { module: shaderModule },
-    fragment: {
-      module: shaderModule,
-      targets: [{ format }],
-    },
-    primitive: { topology: 'triangle-strip' },
-  });
+  const id = `${format}:${code}`;
+  const pipelines = s_deviceToDrawTexturePipelines.get(device) || new Map();
```

Comment:
u-nit: `||`  is used here but s_deviceToDrawTextureShaderModules uses `??`. Can use `??` for both